File: plugins/oura-mcp/catalog.py
"""The Oura API collections exposed by the two-tool MCP surface."""

from __future__ import annotations

from dataclasses import asdict, dataclass
from enum import Enum
from typing import Literal


class RangeKind(str, Enum):
    NONE = "none"
    DATE = "date"
    DATETIME = "datetime"


@dataclass(frozen=True, slots=True)
class CollectionSpec:
    name: str
    family: str
    path: str
    range_kind: RangeKind
    documented_oauth_scopes: tuple[str, ...]
    supports_cursor: bool = True
    supports_latest: bool = False
    notes: str | None = None

    def public_dict(self) -> dict[str, object]:
        value = asdict(self)
        value["range_kind"] = self.range_kind.value
        return value


OuraCollectionName = Literal[
    "daily_activity",
    "daily_cardiovascular_age",
    "daily_readiness",
    "daily_resilience",
    "daily_sleep",
    "daily_spo2",
    "daily_stress",
    "enhanced_tag",
    "heartrate",
    "personal_info",
    "rest_mode_period",
    "ring_battery_level",
    "ring_configuration",
    "session",
    "sleep",
    "sleep_time",
    "tag",
    "vo2_max",
    "workout",
]


_DASHBOARD_SCOPE_NOTE = (
    "Granted by the Oura dashboard; exact public OAuth scope name unresolved."
)

_COLLECTIONS = (
    CollectionSpec(
        "daily_activity", "daily", "daily_activity", RangeKind.DATE, ("daily",)
    ),
    CollectionSpec(
        "daily_cardiovascular_age",
        "heart_health",
        "daily_cardiovascular_age",
        RangeKind.DATE,
        (),
        notes=_DASHBOARD_SCOPE_NOTE,
    ),
    CollectionSpec(
        "daily_readiness", "daily", "daily_readiness", RangeKind.DATE, ("daily",)
    ),
    CollectionSpec(
        "daily_resilience",
        "stress",
        "daily_resilience",
        RangeKind.DATE,
        (),
        notes=_DASHBOARD_SCOPE_NOTE,
    ),
    CollectionSpec("daily_sleep", "daily", "daily_sleep", RangeKind.DATE, ("daily",)),
    CollectionSpec("daily_spo2", "spo2", "daily_spo2", RangeKind.DATE, ("spo2Daily",)),
    CollectionSpec(
        "daily_stress",
        "stress",
        "daily_stress",
        RangeKind.DATE,
        (),
        notes=_DASHBOARD_SCOPE_NOTE,
    ),
    CollectionSpec("enhanced_tag", "tag", "enhanced_tag", RangeKind.DATE, ("tag",)),
    CollectionSpec(
        "heartrate",
        "heartrate",
        "heartrate",
        RangeKind.DATETIME,
        ("heartrate",),
        supports_latest=True,
    ),
    CollectionSpec(
        "personal_info",
        "personal",
        "personal_info",
        RangeKind.NONE,
        ("personal", "email"),
        supports_cursor=False,
    ),
    CollectionSpec(
        "rest_mode_period", "daily", "rest_mode_period", RangeKind.DATE, ("daily",)
    ),
    CollectionSpec(
        "ring_battery_level",
        "ring_configuration",
        "ring_battery_level",
        RangeKind.DATETIME,
        (),
        supports_latest=True,
        notes=_DASHBOARD_SCOPE_NOTE,
    ),
    CollectionSpec(
        "ring_configuration",
        "ring_configuration",
        "ring_configuration",
        RangeKind.NONE,
        (),
        notes=_DASHBOARD_SCOPE_NOTE,
    ),
    CollectionSpec("session", "session", "session", RangeKind.DATE, ("session",)),
    CollectionSpec("sleep", "daily", "sleep", RangeKind.DATE, ("daily",)),
    CollectionSpec("sleep_time", "daily", "sleep_time", RangeKind.DATE, ("daily",)),
    CollectionSpec("tag", "tag", "tag", RangeKind.DATE, ("tag",)),
    CollectionSpec(
        "vo2_max",
        "heart_health",
        "vO2_max",
        RangeKind.DATE,
        (),
        notes=_DASHBOARD_SCOPE_NOTE,
    ),
    CollectionSpec("workout", "workout", "workout", RangeKind.DATE, ("workout",)),
)

COLLECTIONS = {collection.name: collection for collection in _COLLECTIONS}


def collection_spec(name: str) -> CollectionSpec:
    try:
        return COLLECTIONS[name]
    except KeyError as error:
        raise ValueError(f"Unknown Oura collection: {name}") from error


def catalog_payload() -> dict[str, object]:
    return {
        "provider": "oura",
        "api_version": "v2",
        "backend": "synthetic_fixture",
        "collections": [collection.public_dict() for collection in _COLLECTIONS],
        "unresolved": ["nighttime_movement_trace"],
    }

File: plugins/oura-mcp/src/catalog.ts
export type RangeKind = "none" | "date" | "datetime";

export type CollectionSpec = {
  name: string;
  family: string;
  path: string;
  rangeKind: RangeKind;
  documentedScopes: string[];
  supportsCursor: boolean;
  supportsLatest: boolean;
  notes?: string;
};

const dashboardScopeNote =
  "Granted by the Oura developer dashboard; the public OAuth scope table does not name the exact scope string.";

const collectionDefinitions: Array<Omit<CollectionSpec, "supportsCursor" | "supportsLatest"> & Partial<Pick<CollectionSpec, "supportsCursor" | "supportsLatest">>> = [
  { name: "daily_activity", family: "daily", path: "daily_activity", rangeKind: "date", documentedScopes: ["daily"] },
  { name: "daily_cardiovascular_age", family: "heart_health", path: "daily_cardiovascular_age", rangeKind: "date", documentedScopes: [], notes: dashboardScopeNote },
  { name: "daily_readiness", family: "daily", path: "daily_readiness", rangeKind: "date", documentedScopes: ["daily"] },
  { name: "daily_resilience", family: "stress", path: "daily_resilience", rangeKind: "date", documentedScopes: [], notes: dashboardScopeNote },
  { name: "daily_sleep", family: "daily", path: "daily_sleep", rangeKind: "date", documentedScopes: ["daily"] },
  { name: "daily_spo2", family: "spo2", path: "daily_spo2", rangeKind: "date", documentedScopes: ["spo2Daily"] },
  { name: "daily_stress", family: "stress", path: "daily_stress", rangeKind: "date", documentedScopes: [], notes: dashboardScopeNote },
  { name: "enhanced_tag", family: "tag", path: "enhanced_tag", rangeKind: "date", documentedScopes: ["tag"] },
  { name: "heartrate", family: "heartrate", path: "heartrate", rangeKind: "datetime", documentedScopes: ["heartrate"], supportsLatest: true },
  { name: "personal_info", family: "personal", path: "personal_info", rangeKind: "none", documentedScopes: ["personal", "email"], supportsCursor: false },
  { name: "rest_mode_period", family: "daily", path: "rest_mode_period", rangeKind: "date", documentedScopes: ["daily"] },
  { name: "ring_battery_level", family: "ring_configuration", path: "ring_battery_level", rangeKind: "datetime", documentedScopes: [], supportsLatest: true, notes: dashboardScopeNote },
  { name: "ring_configuration", family: "ring_configuration", path: "ring_configuration", rangeKind: "none", documentedScopes: [], notes: dashboardScopeNote },
  { name: "session", family: "session", path: "session", rangeKind: "date", documentedScopes: ["session"] },
  { name: "sleep", family: "daily", path: "sleep", rangeKind: "date", documentedScopes: ["daily"] },
  { name: "sleep_time", family: "daily", path: "sleep_time", rangeKind: "date", documentedScopes: ["daily"] },
  { name: "tag", family: "tag", path: "tag", rangeKind: "date", documentedScopes: ["tag"] },
  { name: "vo2_max", family: "heart_health", path: "vO2_max", rangeKind: "date", documentedScopes: [], notes: dashboardScopeNote },
  { name: "workout", family: "workout", path: "workout", rangeKind: "date", documentedScopes: ["workout"] }
];

export const collections: CollectionSpec[] = collectionDefinitions.map((collection) => ({
  supportsCursor: true,
  supportsLatest: false,
  ...collection
}));

export const collectionNames = collections.map((collection) => collection.name) as [string, ...string[]];

export function collectionSpec(name: string): CollectionSpec {
  const collection = collections.find((candidate) => candidate.name === name);
  if (!collection) throw new Error(`Unknown Oura collection: ${name}`);
  return collection;
}

export function catalogPayload() {
  return {
    provider: "oura",
    api_version: "v2",
    collections,
    unresolved: [
      {
        name: "nighttime_movement_trace",
        status: "not_documented_as_an_api_v2_collection"
      }
    ]
  };
}
